Add a pattern for using a shared object as a cache (#32187)
apache/beam: website/www/site/content/en/documentation/patterns/shared-class.md

---
title: "Cache data using a shared object"
---

<!--
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

# Cache data using a shared object

> A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than stateful `DoFn`.

Reviewer: ```suggestion
A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than a stateful `DoFn`.
```

Nit

Author: Resolved by a separate PR.

> A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than stateful `DoFn`.

Reviewer 2: ```suggestion
A cache is a software component that stores data so that future requests for that data can be served faster. To access a cache, you can use side inputs, stateful `DoFn`, and calling an external service. The Python SDK provides another option in the shared module. This option can be more memory-efficient than side inputs and simpler than a stateful `DoFn`.
```

Author: Resolved by a separate commit.

> The samples on this page show you how to use the Shared class of the [shared module](https://beam.apache.org/releases/pydoc/current/apache_beam.utils.shared.html) to enrich elements in both bounded and unbounded `PCollection`s. Two data sets are used in the samples - order and customer. The order records include customer IDs with which customer attributes are added by mapping the customer records.

Reviewer 2: ```suggestion
The examples on this page demonstrate how to use the `Shared` class of the [`shared module`](https://beam.apache.org/releases/pydoc/current/apache_beam.utils.shared.html) to enrich elements in both bounded and unbounded `PCollection` objects. Two data sets are used in the samples: _order_ and _customer_. The order records include customer IDs that customer attributes are added to by mapping the customer records.
```

Author: Resolved by a separate commit.

## Create a cache on a batch pipeline

> In this example, the customer cache is loaded as a dictionary in the `setup` method of the `EnrichOrderFn`, and it is used to add customer attributes to the order records. Note that a Shared object encapsulates a weak reference to a singleton instance of the shared resource. Therefore, we need to create a wrapper class since the Python dictionary does not support weak references.

Reviewer 2: ```suggestion
In this example, the customer cache is loaded as a dictionary in the `setup` method of the `EnrichOrderFn`. The cache is used to add customer attributes to the order records. Because the Python dictionary doesn't support weak references and a `Shared` object encapsulates a weak reference to a singleton instance of the shared resource, create a wrapper class.
```

Author: Resolved by a separate commit.

> {{< highlight py >}}

Reviewer: I think there are a number of pieces we don't need here for our example, and we could greatly simplify by just providing an implementation of `EnrichOrderFn`. So this example would become:

```
class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._version = 1
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers(version=self._version)
        return WeakRefDict(self._customers)

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

Author: Let me update with the following snippet in a separate commit.

```python
class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers()
        return WeakRefDict(self._customers)

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

Author: Resolved by a separate commit.

import random
import logging
import time
from datetime import datetime
from uuid import uuid4

import apache_beam as beam
from apache_beam.utils import shared

def gen_customers(version: int, num_cust: int = 1000):
    d = dict()
    for r in range(num_cust):
        d[r] = {"version": version}
    d["timestamp"] = datetime.now().timestamp()
    return d

def gen_orders(ts: float, num_ord: int = 5, num_cust: int = 1000):
    orders = [
        {
            "order_id": str(uuid4()),
            "customer_id": random.randrange(1, num_cust),
            "timestamp": ts,
        }
        for _ in range(num_ord)
    ]
    for o in orders:
        yield o

> # wrapper class needed for a dictionary since it does not support weak references.

Reviewer 2: ```suggestion
# The wrapper class is needed for a dictionary, because it does not support weak references.
```

Author: Resolved by a separate commit.

class WeakRefDict(dict):
    pass

class EnrichOrderFn(beam.DoFn):
    def __init__(self, shared_handle):
        self._version = 1
        self._customers = {}
        self._shared_handle = shared_handle

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers)

    def load_customers(self):
        time.sleep(2)
        self._customers = gen_customers(version=self._version)
        return WeakRefDict(self._customers)

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}

with beam.Pipeline() as p:
    shared_handle = shared.Shared()
    (
        p
        | beam.Create(gen_orders(ts=datetime.now().timestamp()))
        | beam.ParDo(EnrichOrderFn(shared_handle))
        | beam.Map(print)
    )

    logging.getLogger().setLevel(logging.INFO)
    logging.info("Building pipeline ...")
{{< /highlight >}}

## Create a cache and update it regularly on a streaming pipeline

> As the batch pipeline example, the customer cache is initialized as a dictionary in the `setup` method. To update the cache, it is refreshed when its timestamp is earlier than the defined threshold value of 5 seconds. The refreshment is implemented in the `start_bundle` method. Once the cached is updated, the order records are enriched with updated customer attributes. Note that it is necessary to change the `tag` argument of the `acquire` method to reload the shared object when it gets refreshed. The current timestamp value (`current_ts`) is used in this example.

Reviewer 2: ```suggestion
The customer cache is initialized as a dictionary in the `setup` method. To update the cache, it is refreshed when its timestamp is earlier than the defined threshold value of 5 seconds. The refresh is implemented in the `start_bundle` method. After the cached is updated, the order records are enriched with updated customer attributes. To reload the shared object when it gets refreshed, you must change the `tag` argument of the `acquire` method. This example uses the current timestamp value (`current_ts`).
```

Author: Resolved by a separate commit.

{{< highlight py >}}
import random
import logging
import time
from datetime import datetime
from uuid import uuid4

import apache_beam as beam
from apache_beam.utils import shared
from apache_beam.transforms.periodicsequence import PeriodicImpulse

def gen_customers(version: int, num_cust: int = 1000):
    d = dict()
    for r in range(num_cust):
        d[r] = {"version": version}
    d["timestamp"] = datetime.now().timestamp()
    return d

def gen_orders(e: float, num_ord: int = 5, num_cust: int = 1000):
    orders = [
        {
            "order_id": str(uuid4()),
            "customer_id": random.randrange(1, num_cust),
            "timestamp": e,
        }
        for _ in range(num_ord)
    ]
    for o in orders:
        yield o

> # wrapper class needed for a dictionary since it does not support weak references.

Reviewer 2: ```suggestion
# The wrapper class is needed for a dictionary, because it doesn't support weak references.
```

Author: Resolved by a separate commit.

class WeakRefDict(dict):
    pass

class EnrichOrderFn(beam.DoFn):
    def __init__(self, shared_handle):
        self._max_stale_sec = 5
        self._version = 1
        self._customers = {}
        self._shared_handle = shared_handle

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(

>             self.load_customers, datetime.now().timestamp()

Reviewer: Calling acquire with a timestamp based tag in setup is probably a bad idea since this will be consistently changing for each thread which calls setup.

Author: Would you consider it is okay if it is changed as

```python
    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers, "init")
```

Reviewer: I think using the timestamp bucketized version I mentioned below probably makes more sense (see that thread for explanation of why)

Author: Resolved by a separate commit.

        )

    def load_customers(self):
        time.sleep(2)
        self._customers = gen_customers(version=self._version)
        return WeakRefDict(self._customers)

    def start_bundle(self):

>         # update the lookup table when the time difference exceeds the threshold.

Reviewer 2: ```suggestion
        # Update the lookup table when the time difference exceeds the threshold.
```

Author: Resolved by a separate commit.

        current_ts = datetime.now().timestamp()

>         ts_diff = current_ts - self._customers["timestamp"]

Reviewer: Rather than using `ts_diff` which can be different for each thread (causing the cache to be reloaded several times), it would be better to use a more deterministic mapping.

For example, you could do something like:

`tag = current_ts - (current_ts % stale._max_stale_sec)`. This would ensure that you're always still within `max_stale_sec`, but you're not reloading the cache more than needed. There may be an even smarter strategy here using a second shared object (which could tell you, for example, the tag of the latest results), but I think this is probably good enough

Author: Can you clarify more? `ts_diff` is necessary to check whether the shared object should be refreshed or not. The tag value can be updated to be more deterministic. Would you consider the following snippet is okay?

```python
class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._max_stale_sec = 60
        self._cache_version = 0
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers, self._cache_version)
        self._cache_version +=1
 
    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers()
        return WeakRefDict(self._customers)

    def start_bundle(self):
        # Update the lookup table when the time difference exceeds the threshold.
        current_ts = datetime.now().timestamp()
        ts_diff = current_ts - self._customers["timestamp"]
        if ts_diff > self._max_stale_sec:
            self._customer_lookup = self._shared_handle.acquire(self.load_customers, self._version)
            elf._cache_version +=1

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

Reviewer: This is probably a bit better, but it still has some potential issues. Imagine the following sequence of events:

```
Worker spins up DoFn1 and DoFn2
DoFn1.setup() acquires shared_handle with _cache_version = 0
DoFn2.setup() acquires shared_handle with _cache_version = 0 (so far, so good)
DoFn1.start_bundle() acquires shared_handle with _cache_version = 1 (still ok - 2 copies of cache exist, but both are actually being used)
DoFn1.start_bundle() acquires shared_handle with _cache_version = 2 (at this point, _cache_version=1 is garbage collected)
DoFn2.start_bundle acquires shared_handle with _cache_version = 1 (this doesn't make sense, it should just jump to _cache_version=2, but the versions are out of sync)
```

Since there's no guarantee that bundle requests are going to get round-robined, over time there could be drift between the various `DoFn`'s `_cache_version`s, leading to extra remote calls. Over time, it is reasonably likely that the cache will degrade to the point where we're not getting any benefit from it.

That's why I suggested still grounding the current `_cache_version` in the current time, but bucketizing it so that it still resolves to a single discrete value per bucket

Author: Thanks for the clarification. Would you consider the following change is acceptable?

```python
# The wrapper class is needed for a dictionary, because it does not support weak references.
class WeakRefDict(dict):
    pass

class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._max_stale_sec = 60
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(
            self.load_customers, self.set_tag()
        )
 
    def set_tag(self):
        current_ts = datetime.now().timestamp()
        return current_ts - (current_ts % self._max_stale_sec)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers(tag=self.set_tag())
        return WeakRefDict(self._customers)

    def start_bundle(self):
        # Update the lookup table when the time difference exceeds the threshold.
        if self.set_tag() > self._customers["tag"]:
            self._customer_lookup = self._shared_handle.acquire(
                self.load_customers, self.set_tag()
            )

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

Author: Resolved by a separate commit.

---
title: "Cache data using a shared object"
---

<!--
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

# Cache data using a shared object

> A cache is a software component that stores data so that future requests for that data can be served faster. To access a cache, you can use side inputs, stateful `DoFn`, and calling an external service. The Python SDK provides another option in the shared module. This option can be more memory-efficient than side inputs, simpler than a stateful `DoFn`, and more performant than calling an external service because it does not have to access an external service for every element or bundle of elements. See [this session](https://2022.beamsummit.org/sessions/strategies-for-caching-data-in-dataflow-using-beam-sdk/) at the Beam Summit 2022 for more details about strategies for caching data using Beam SDK.

Reviewer 2: ```suggestion
A cache is a software component that stores data so that future requests for that data can be served faster. To access a cache, you can use side inputs, stateful `DoFn`, and calls to an external service. The Python SDK provides another option in the shared module. This option can be more memory-efficient than side inputs, simpler than a stateful `DoFn`, and more performant than calling an external service, because it does not have to access an external service for every element or bundle of elements. For more details about strategies for caching data using Beam SDK, see the session [Strategies for caching data in Dataflow using Beam SDK](https://2022.beamsummit.org/sessions/strategies-for-caching-data-in-dataflow-using-beam-sdk/) from the 2022 Beam Summit.
```

In this example, the customer cache is loaded as a dictionary in the `setup` method of the `EnrichOrderFn`. The cache is used to add customer attributes to the order records. Because the Python dictionary doesn't support weak references and a `Shared` object encapsulates a weak reference to a singleton instance of the shared resource, create a wrapper class.

{{< highlight py >}}
# The wrapper class is needed for a dictionary, because it does not support weak references.
class WeakRefDict(dict):
    pass

class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers()
        return WeakRefDict(self._customers)

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
{{< /highlight >}}

## Create a cache and update it regularly on a streaming pipeline

> The customer cache is assumed to change over time, and it should be refreshed periodically. Note that, to reload the shared object, you must change the `tag` argument of the `acquire` method. In this example, the refresh is implemented in the `start_bundle` method where it compares the current tag value to the value that is associated with the existing shared object. The `set_tag` method returns a tag value that is the same within the maximum seconds of staleness. Therefore, if a tag value is greater than the existing tag value, it triggers to refresh the customer cache.

Reviewer 2: ```suggestion
Because the customer cache is assumed to change over time, you need to refresh it periodically. To reload the shared object, change the `tag` argument of the `acquire` method. In this example, the refresh is implemented in the `start_bundle` method, where it compares the current tag value to the value that is associated with the existing shared object. The `set_tag` method returns a tag value that is the same within the maximum seconds of staleness. Therefore, if a tag value is greater than the existing tag value, it triggers a refresh of the customer cache.
```
